Snapshot access mode on transaction start

henriknyman opened the pull request:
A snapshot of the access mode is taken on begin transaction, which will keep
the authorization state at that point active for the life time of the transaction.

This fixes the issue that authorization can expire during long running
transactions with LDAP authorization
(with `dbms.security.realms.ldap.authorization.use_system_account=false`) or
plugin authorization using the simplified `AuthPlugin` interface.

changelog: Keep a snapshot of the authorization state for the life time of a transaction

henriknyman committed:
Snapshot access mode on transaction start

A snapshot of the access mode is taken on begin transaction, which will keep
the authorization state at that point active for the life time of the transaction.

This fixes the issue that authorization can expire during long running
transactions with LDAP authorization
(with `dbms.security.realms.ldap.authorization.use_system_account=false`) or
plugin authorization using the simplified `AuthPlugin` interface.

henriknyman committed:
Fix some failing tests

- Fix username in list queries
- Fix enterprise auth procedures requiring an admin
- Remove snapshot method on AccessMode, since we always need
an instance of AccessModeSnapshot for DbmsOperations to acquire
the original AuthSubject

henriknyman committed:
Fix bug in bolt protocol on connection fatality

On Bolt connection fatality, the failure message may not be sent to the client
if there are more message waiting to be processed.

This adds onFatal() to BoltResponseMessageHandler, that does the same as
onFailure() but also flushes the output, since we know that the connection
will be immediately closed from the server.

henriknyman committed:
Fix more tests

- Plugin authorization expired
- Move assertion for test session killed into interaction level,
and use eventuallyDisconnects() on Bolt

henriknyman committed:
Fix procedure call with allowed

Procedure calls with the `allowed` annotation should also handle
`AccessModeSnapshot`.

henriknyman committed:
Fix license header and code style

henriknyman committed:
Add simple test cases

Should not get auth expired with Ldap or Plugin if cache expires within
a transaction.

henriknyman committed:
Handle recursive overridden AccessMode

- Move getOriginalAccessMode() to AccessMode base class
- Move username() to AccessMode base class to get rid of static downcast
method. When we have a proper security context we can remove it again.

henriknyman committed:
Reintroduce getSnapshot() on AccessMode

- Default subclass implementations creates AccessModeSnapshot
- Static access modes and snapshot return `this`
- Some cleanup of interface code

henriknyman committed:
Support snapshot of `allowsProcedureWith`

- Add support to snapshot `AuthorizationInfo` in the enterprise auth manager,
- Add tests for nested procedure calls with allowed annotation

henriknyman committed:
Move `allowsProcedureWith` to `AccessMode`

- Rename AccessModeSnapshot create method

henriknyman committed:
Fix nested allowed procedure call

A nested procedure call with allowed should not be able to elevate the
access mode above the original elevated access mode. We do this by only
allowing one level of access mode where overrideOriginalMode() is true.

henriknyman committed:
Refactor OverriddenAccessMode

Separate it into two implementation classes to clarify the difference between
restricting and elevating access mode,

- Replace OverriddenAccessMode with abstract LayeredAccessMode
- Separate implementation classes RestrictedAccessMode and ElevatedAccessMode
- Rename overrideOriginalMode() to isElevated(), with default implementation in
base interface.

Also fixes function call allowed, which had a bug of not reverting overridden
access mode

Mats-SX commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/security/OverriddenAccessMode.java:
```diff
@@ -80,6 +80,12 @@ public String name()
+    @Override
+    public AccessMode getSnapshot()
+    {
+        return null;
```
Shouldn't this be `overridden.getSnapshot()`?

Mats-SX replied:
It just feels a bit unfriendly to return `null`.

henriknyman replied:
This has already been changed in refactoring

Mats-SX commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v3_1/TransactionBoundQueryContext.scala:
```diff
@@ -586,36 +585,37 @@ final class TransactionBoundQueryContext(val transactionalContext: Transactional
+
+  private def shouldElevate(allowed: Array[String]): Boolean = {
+    // We have to be careful with elevation, since we cannot elevate permissions in a nested procedure call
+    // above the original allowed procedure mode. With enforce this by checking if mode is already an overridden mode.
```
Typo: `With` -> `We`

henriknyman replied:
Fixed

henriknyman committed:
Add more tests

Fix correct restriction of allowed procedures nested in regular procedures
- Change isElevated to isOverridden
- Add tests for nested user defined functions
- Add more complex tests for nested procedures

henriknyman committed:
Fix code style violation

Mats-SX commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v3_1/TransactionBoundQueryContext.scala:
```diff
@@ -586,34 +585,39 @@ final class TransactionBoundQueryContext(val transactionalContext: Transactional
+    // We have to be careful with elevation, since we cannot elevate permissions in a nested procedure call
+    // above the original allowed procedure mode. We enforce this by checking if mode is already an overridden mode.
+    val mode = transactionalContext.accessMode
+    allowed.nonEmpty && !mode.isOverridden && mode.getOriginalAccessMode.allowsProcedureWith(allowed)
```
How about the case where a `WRITE` procedure calls a `READ` procedure? That's now in the domain of `override` rather than `elevate`, is that correct?

henriknyman replied:
Yes, we only elevate access on the outermost procedure call. Once you got your elevated access rights, they can only be restricted by nested procedure calls. In the case of a WRITE procedure calling a READ procedure, it will get restricted to READ in the nested call. To be able to call the outer WRITE procedure you either need WRITE permissions or to be allowed WRITE permissions by annotation, which then implies READ permissions. Since the procedure modes have a strict order of which mode implies another, this should be sufficient.

Mats-SX commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v3_1/TransactionBoundQueryContext.scala:
```diff
@@ -586,34 +585,39 @@ final class TransactionBoundQueryContext(val transactionalContext: Transactional
+    val call: KernelProcedureCall =
+      if (shouldElevate(allowed))
+        transactionalContext.statement.procedureCallOperations.procedureCallSchemaElevated(_, _)
+      else
```
Nice cleanup.

Mats-SX commented on community/kernel/src/main/java/org/neo4j/kernel/api/ProcedureCallOperations.java:
```diff
@@ -82,14 +79,28 @@
+     * @param arguments the function arguments.
+     * @throws ProcedureException if there was an exception thrown during function execution.
+     */
+    Object functionCallElevated( QualifiedName name, Object[] arguments ) throws ProcedureException;
```
Still not convinced on this one :)

Mats-SX commented on community/kernel/src/main/java/org/neo4j/kernel/api/ProcedureCallOperations.java:
```diff
@@ -82,14 +79,28 @@
+     * @param arguments the function arguments.
+     * @throws ProcedureException if there was an exception thrown during function execution.
+     */
+    Object functionCall( QualifiedName name, Object[] arguments ) throws ProcedureException;
```
I guess that wasn't introduced by this PR, but should we really throw `ProcedureException` here?

MishaDemianenko replied:
This interface javadoc describe current interface as: `Specifies procedure call operations for the three types of procedure calls that can be made.`
why functions was moved here? if we distinguish them they should not be part of this class and corresponding interface

Mats-SX commented on community/kernel/src/main/java/org/neo4j/kernel/api/security/AccessMode.java:
```diff
@@ -228,142 +175,55 @@ public boolean allowsSchemaWrites()
+     * @return
+     * @throws InvalidArgumentsException
+     */
+    default boolean allowsProcedureWith( String[] roleNames ) throws InvalidArgumentsException
```
This sneakily introduces the concept of `role` into community. Let's perhaps rename this parameter to `allowed`?

Mats-SX commented on community/kernel/src/main/java/org/neo4j/kernel/api/security/AccessMode.java:
```diff
@@ -228,142 +175,55 @@ public boolean allowsSchemaWrites()
+
+    /**
+     * Determines whether this subject is allowed to execute a procedure with the parameter string in its
+     * procedure annotation.
```
'this subject' is incorrect in this context. `this` is an `AccessMode`, which is different from a subject. I suggest something like this:

Determines whether this mode allows execution of a procedure with the parameter string array in its procedure annotation.

Mats-SX commented on community/kernel/src/main/java/org/neo4j/kernel/api/security/AccessMode.java:
```diff
@@ -228,142 +175,55 @@ public boolean allowsSchemaWrites()
+     *
+     * @param roleNames
+     * @return
+     * @throws InvalidArgumentsException
```
I'd like a JavaDoc for these three as well. Specifically `return` would be helpful. I'm thinking something along the lines of 

`true` if (and only if?) the parameter array contains a string that encodes a privilege to allow this mode to execute it

but it feels a bit contrived. Is this really the correct place for this method?

Mats-SX commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelStatement.java:
```diff
@@ -225,8 +223,8 @@ final void forceClose()
-        AccessMode mode = transaction.mode();
-        return ( mode instanceof AuthSubject ) ? Optional.of( ((AuthSubject) mode).username() ) : Optional.empty();
+        String username = transaction.mode().username();
+        return Optional.of( username );
```
Could inline.

Mats-SX commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java:
```diff
@@ -800,7 +799,7 @@ public long getCommitTime()
     {
         AccessMode oldMode = this.accessMode;
-        this.accessMode = new OverriddenAccessMode( oldMode, mode );
+        this.accessMode = mode;
```
Nice simplification.

Mats-SX commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/security/RestrictedAccessMode.java:
```diff
@@ -0,0 +1,55 @@
+public class RestrictedAccessMode extends LayeredAccessMode
+{
+    public RestrictedAccessMode( AccessMode originalMode,
+            AccessMode overriddenAccessMode )
```
It is a bit confusing reading these as `original` and `overridden`. How about `original` and `restricting`? Or `restrictedTo`?

Mats-SX commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/security/ElevatedAccessMode.java:
```diff
@@ -0,0 +1,55 @@
+public class ElevatedAccessMode extends LayeredAccessMode
+{
+    public ElevatedAccessMode( AccessMode originalMode,
+            AccessMode overriddenAccessMode )
```
Could we call these `original` and `elevating` or `elevatedTo`?

Mats-SX commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/security/LayeredAccessMode.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    @Override
+    public boolean isOverridden()
```
Isn't it a bit strange to find that `ElevatedAccessMode.isOverridden()` exists and is true? It's not overridden, really.

henriknyman replied:
No, the meaning of `isOverridden()` is that it is either restricted or elevated, i.e. you have an access mode that overrides the original access mode.

Mats-SX commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/security/LayeredAccessMode.java:
```diff
@@ -0,0 +1,65 @@
+    public LayeredAccessMode( AccessMode originalMode, AccessMode overriddenMode )
+    {
+        this.originalMode = originalMode;
+        this.overriddenMode = overriddenMode;
```
I guess the field naming comments apply the most to this class. Not sure what a good common word is for the two subclasses though. Perhaps one of `layered`, `additional`, `other`?

henriknyman replied:
overridden here just means either restricted or elevated, it is overriding the original access mode in some way.

Mats-SX commented on enterprise/security/src/test/java/org/neo4j/server/security/enterprise/auth/AuthScenariosInteractionTestBase.java:
```diff
@@ -536,7 +482,7 @@ public void customRoleWithProcedureAccess() throws Exception
         testFailTestProcs( mats );
         assertEmpty( adminSubject, "CALL dbms.security.addRoleToUser('role1', 'mats')" );
-        testSuccessfulTestProcs( mats );
+        //testSuccessfulTestProcs( mats ); // TODO: FIXME
```
Leftover TODO

henriknyman replied:
We should fix this

Mats-SX commented on enterprise/security/src/test/java/org/neo4j/server/security/enterprise/auth/EmbeddedInteraction.java:
```diff
@@ -170,6 +171,13 @@ public void assertInitFailed( EnterpriseAuthSubject subject )
+    public void assertSessionKilled( EnterpriseAuthSubject subject )
+    {
+        assertThat( executeQuery( subject, "MATCH (n:Node) RETURN count(n)", null, r -> {} ),
+                containsString( "Read operations are not allowed" ) );
```
Why is this parametrised with a subject? This assertion will give a false negative (as in: not throwing when it should) if the parameter subject doesn't have read privileges, right?

henriknyman replied:
The subject is used in `BoltInteraction` to get the transport connection. The asserts in REST and embedded doesn't make much sense since there is no session that could be killed. Instead they basically check that something happened so that the subject now doesn't have any permissions (e.g. user was deleted or suspended). This is an old remnant of the unifying of the tests across different surfaces.

Mats-SX replied:
Makes sense.

Mats-SX commented on enterprise/security/src/test/java/org/neo4j/server/security/enterprise/auth/ProcedureInteractionTestBase.java:
```diff
@@ -559,28 +543,67 @@ TransportConnection startBoltSession( String username, String password ) throws
         }
 
-        @Procedure( name = "test.allowedProcedure1", allowed = {"role1"}, mode = Mode.READ )
+        @Procedure( name = "test.allowedReadProcedure", allowed = {"role1"}, mode = Mode.READ )
```
Nice that you renamed these.

Mats-SX commented on enterprise/security/src/test/java/org/neo4j/server/security/enterprise/auth/integration/bolt/LdapAuthenticationIT.java:
```diff
@@ -323,15 +321,38 @@ public void shouldFailIfAuthorizationExpiredWithUserLdapContext() throws Throwab
         // Then
         assertAuth( "neo4j", "abc123" );
 
+        // We need to replace this with a two thread test.
```
Should we have a card for it, or fix it on this PR?

henriknyman replied:
It was just a leftover obsolete comment that should be removed

Mats-SX reviewed (changes requested):
Lots of comments, but most are about naming. I like the refactoring into `LayeredAccessMode`, but it has some fields/methods that may need amending for clarification of the new purpose.

henriknyman committed:
Fix test having commented-out assertion

henriknyman committed:
Fix review comments

henriknyman committed:
Fix user defined function call permission check

MishaDemianenko commented on community/bolt/src/main/java/org/neo4j/bolt/v1/messaging/BoltResponseMessageHandler.java:
```diff
@@ -41,4 +41,5 @@
 
     void onFailure( Status status, String message ) throws E;
 
+    default void onFatal( Status status, String message ) throws E { onFailure( status, message ); }
```
I find this super confusing: it's default implementation that will delegate to `onFailure` handler silently while in the code we try to do distinction:

```
            if ( error.isFatal() )
            {
                out.onFatal( error.status(), error.message() );
            }
            else
            {
                out.onFailure( error.status(), error.message() );
            }
```

that kind of have the idea that we will it's important to distinguish those cases

henriknyman replied:
The default implementation was just to avoid having to modify a lot of test files. But in this case I guess it makes sense to remove the default. In the other classes below I would prefer to keep the default implementations to keep the change set down.

MishaDemianenko commented on community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/Neo4jError.java:
```diff
@@ -148,29 +150,53 @@ public static Status codeFromString( String codeStr )
             if (cause instanceof OutOfMemoryError)
             {
-                return new Neo4jError( Status.General.OutOfMemoryError, cause );
+                return new Neo4jError( Status.General.OutOfMemoryError, cause, isFatal );
```
In case of `OutOfMemoryError` or `StackOverflowError` we will pass `cause` as cause instead of source `any`. Why it's like that? Would it make sense to always see full exception?

henriknyman replied:
I don't know

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/api/security/AccessMode.java:
```diff
@@ -228,142 +175,56 @@ public boolean allowsSchemaWrites()
     AuthorizationViolationException onViolation( String msg );
     String name();
+
+    default String username()
```
I think we missing some javadocs here: how access mode is connected with username and what username we expecting here

Mats-SX replied:
We plan on separating the concepts of AuthSubject and AccessMode soon, so we rather make the docs effort then.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/api/security/AccessMode.java:
```diff
@@ -228,142 +175,56 @@ public boolean allowsSchemaWrites()
+        return ""; // Should never clash with a valid username
+    }
+
+    default AccessMode getOriginalAccessMode()
```
i find weird combinations of default and non default methods that implemented in 
accessMode and Static. I would suggest we remove all the defaults since that will force users override required methods instead of spend hours in debug

Mats-SX replied:
Nearly all implementations will do `return this` here. On this instance, we like the code sharing.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/api/ProcedureCallOperations.java:
```diff
@@ -40,15 +39,14 @@
 
     /**
-     * Invoke a read-only procedure by name, and override the transaction's access mode with
+     * Invoke a read-only procedure by name, and elevate the transaction's access mode to
```
All java docs claim that access mode will be elevated to the given and given is gone now

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/dbms/NonTransactionalDbmsOperations.java:
```diff
@@ -47,9 +47,11 @@ public NonTransactionalDbmsOperations( Procedures procedures )
         BasicContext ctx = new BasicContext();
-        if ( mode instanceof AuthSubject )
+
+        AccessMode originalMode = mode.getOriginalAccessMode();
```
why do we use original mode here? what if we have situation:
original mode (full) -elevated-> mode(read) -try to elevate -> mode(write).
Original mode will allow it, but that mode was restricted by nested procedure. And now procedure nested in nested can actually get more higher access mode then outer.

henriknyman replied:
We are only using it to get hold of the `AuthSubject` through a downcast. The only permission checks done in dbms procedures are done directly on auth subject, like `isAdmin()` or has a username.
We have a refactoring planned of separating access mode and auth subject into a security context, but right now you get auth subject through the reference of the access mode.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/security/AccessModeSnapshot.java:
```diff
@@ -0,0 +1,103 @@
+    {
+        // TODO: Use flyweight pattern instead of always creating new objects, when we have a proper
+        //       security context and do not need to obtain the original mode through this object
+        return new AccessModeSnapshot( accessMode );
```
is it not the case already now that our accessMode objects are final and can't be modified?

henriknyman replied:
No, because the access mode could be an `AuthSubject` (it almost always is), and we need to be able to get hold of it by a downcast of the original access mode (until we can do the refactoring mentioned above). I guess we could still do a flyweight including the original access mode reference, but it would have much more combinations so it didn't seem worthwhile.

fickludd replied:
well no, in the current architecture the `StandardEnterpriseAuthSubject` (which extends AccessMode) dynamically checks permissions on each allows\* call. This PR is about freezing these permissions in a snapshot on transaction start.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/security/ElevatedAccessMode.java:
```diff
@@ -0,0 +1,55 @@
+
+public class ElevatedAccessMode extends LayeredAccessMode
+{
+    public ElevatedAccessMode( AccessMode originalMode,
```
since we elevate access mode to a new one should be the case that we should only use elevatedMode to check what we can do now?

Mats-SX replied:
Good catch! Fixed.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/security/ElevatedAccessMode.java:
```diff
@@ -0,0 +1,55 @@
+
+import org.neo4j.kernel.api.security.AccessMode;
+
+public class ElevatedAccessMode extends LayeredAccessMode
```
Javadocs missing

Mats-SX replied:
Fixed!

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/security/LayeredAccessMode.java:
```diff
@@ -0,0 +1,65 @@
+import org.neo4j.graphdb.security.AuthorizationViolationException;
+import org.neo4j.kernel.api.security.AccessMode;
+
+abstract class LayeredAccessMode implements AccessMode
```
java docs is missing; what layers are we talking about here?

Mats-SX replied:
Renamed and fixed.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/security/RestrictedAccessMode.java:
```diff
@@ -0,0 +1,55 @@
+
+public class RestrictedAccessMode extends LayeredAccessMode
+{
+    public RestrictedAccessMode( AccessMode originalMode,
```
have the same question as above here if we restrict our mode to some restrictive mode should we just simply use that mode as a source of mode truth?

Mats-SX replied:
Given that we do explicit checks on the original mode before overriding, yes. Change made.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelStatement.java:
```diff
@@ -225,8 +223,8 @@ final void forceClose()
     {
-        AccessMode mode = transaction.mode();
-        return ( mode instanceof AuthSubject ) ? Optional.of( ((AuthSubject) mode).username() ) : Optional.empty();
+        String username = transaction.mode().username();
```
since mode is always present now, does it make sense still to have Optional result?

Mats-SX replied:
Yes, you're right. Changed.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java:
```diff
@@ -230,7 +229,7 @@ public KernelTransactionImplementation initialize(
         this.transactionEvent = tracer.beginTransaction();
         assert transactionEvent != null : "transactionEvent was null!";
-        this.accessMode = accessMode;
+        this.accessMode = accessMode.getSnapshot();
```
the same question as before: accessMode should be non modifiable why do we need to snapshot it?

Mats-SX replied:
See above; it is not immutable generally.

MishaDemianenko commented on enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/MultiRealmAuthManager.java:
```diff
@@ -221,4 +224,22 @@ public void clearAuthCache()
+                        .getAuthorizationInfoSnapshot( principalCollection );
+                if ( info != null )
+                {
+                    infoList.add( info );
```
If we really expect only one instance of AuthorizationInfo to be added into the list we should stop iteration over loop here, otherwise 1 in list constructor is redundant

MishaDemianenko commented on enterprise/security/src/test/java/org/neo4j/server/security/enterprise/auth/ProcedureInteractionTestBase.java:
```diff
@@ -357,8 +335,14 @@ void assertPasswordChangeWhenPasswordChangeRequired( S subject, String newPasswo
         String err = assertCallEmpty( subject, call );
-        assertThat( err, not( equalTo( "" ) ) );
-        assertThat( err, containsString( partOfErrorMsg ) );
+        if ( partOfErrorMsg == null || partOfErrorMsg == "" )
```
StringUtils.isEmpty( partOfErrorMsg )

henriknyman replied:
Hmm, yes we had that first, but encountered some weird runtime dependency issue (that I cannot remember exactly what it was now), and didn't feel we wanted to spend more time on it.

MishaDemianenko commented on enterprise/security/src/test/java/org/neo4j/server/security/enterprise/auth/ProcedureInteractionTestBase.java:
```diff
@@ -357,8 +335,14 @@ void assertPasswordChangeWhenPasswordChangeRequired( S subject, String newPasswo
-        assertThat( err, containsString( partOfErrorMsg ) );
+        if ( partOfErrorMsg == null || partOfErrorMsg == "" )
+        {
+            assertThat( err, not( equalTo( "" ) ) );
```
and since we know it's null or empty here, why just not to use `nullValue` here instead of negation?

MishaDemianenko reviewed (commented):
I finish my review round, i think some of the comments popup just because i'm do not really know how that api should be used and in which way, so they are more questions then actual comments. Looks good in general!

Mats-SX committed:
Fix bug with elevating access mode

- Modify javadocs
- Replace concept of elevated access mode  with overridden access mode

Mats-SX committed:
Make username non-optional

Mats-SX committed:
Use StringUtils.isEmpty()

fickludd committed:
Clarifications and javadocs
